jax/_src/interpreters/mlir.py

```diff
@@ -1463,82 +1437,82 @@ def get_override_lowering_rule(primitive: core.Primitive) -> LoweringRule | None
+                       description: str,
+                       platform_rules: dict[str, LoweringRule],
+                       default_rule: Optional[LoweringRule],
+                       effects: effects_lib.Effects,
+                       *rule_args: ir.Value,
+                       **rule_kwargs) -> ir.Value:
+  """Emits code for a primitive for the current lowering platform(s).
 
   For example, given
-      ctx.module_context.lowering_parameters.platforms = ("cpu", "gpu", "tpu")
+      platform_rules = dict(tpu=rule0, cpu=rule0)
+      default_rule = rule1
+
   and
-      rules = [(["tpu", "cpu"], rule0),
-               (None, rule1)
+      ctx.module_context.lowering_parameters.platforms = ("cpu",)
+
+  emits:
+      rule0(ctx, *rule_args, **rule_kwargs)
+
+  In case of multi-platform lowering, e.g., if
+      ctx.module_context.lowering_parameters.platforms = ("cpu", "cuda", "tpu")
+
   emits:
     rule_idx = case current_platform_idx:
                    0: return 0  # cpu rule index
-                   1: return 1  # gpu rule index
+                   1: return 1  # cuda rule index
                    2: return 0  # tpu rule index
     output = case rule_idx
                0: return rule0(*rule_args, **rule_kwargs)
                1: return rule1(*rule_args, **rule_kwargs)
 
-  If the primitive has a single lowering rule for all platforms of interest,
-  skips the conditionals and emits the same code as for classic single-platform
-  lowering.
-
   Args:
    ctx: lowering context.
    description: a string to include in error messages.
-   rules: a sequence of per-platform rules. Each entry is a tuple, with the
-     first element specifying the platforms, either a sequence of applicable
-     platform names (maybe empty), or None to denote a default entry to use
-     when no other entry applies. The second element of the tuple is a
-     lowering rule, i.e., a function to invoke with a
-     LoweringRuleContext (a sub-context of `ctx`),
-     and `*rule_args` and `**rule_kwargs`.
+   platform_rules: map platform names, e.g., "cpu", "cuda", to
+     `LoweringRule`s, for the platforms that have non-default lowering.
+   default_rule: an optional rule to use for platforms not in `platform_rules`.
+   effects: the set of effects for the current primitive.
    rule_args: the args of the lowering rules.
    rule_kwargs: the kwargs of the lowering rules.
   """
   platforms: Sequence[str] = ctx.module_context.platforms
-  platforms_with_specific_rules: Sequence[str] = util.flatten(
-    [ps for ps, _ in rules if ps is not None])
-  platforms_with_default_rule = [p for p in platforms
-                                 if p not in platforms_with_specific_rules]
-  kept_rules: list[MultiPlatformLoweringRule] = []  # Only the rules for platforms of interest
+  # Special case the common case (single-platform lowering)
+  if len(platforms) == 1:
+    rule = platform_rules.get(platforms[0], default_rule)
+    if rule is None:
```

Comment:
Am I right that this is only possible when `platform_rules` doesn't have a rule for the current platform *and* `default_rule` is not specified?

Reply:
Yes, that would be the only case when this is possible